feat: remove code table

vmx opened the pull request:
Instead of using a code table, derive from a Multihash enum which defines
implementations.

This leads to less generated code.

This PR is *not* intended to be merged as-is. It should be used as a basis for discussions.

I think I found a way to simplify things. I'm not fully done yet (e.g. I think I'd like to split the `MultihashDigest` trait into two pieces), but I thought I push it here to get early feedback if this might be a way to move forward.

I also have no idea how this will work with rust-cid, I haven't looked into that at all yet.

vmx committed:
feat: remove code table

Instead of using a code table, derive from a Multihash enum which defines
implementations.

This leads to less generated code.

dvc94ch commented:
I had a quick look. So I guess multihash returns and error if the constant is out of bounds? I'll review it tomorrow

vmx committed:
feat: split MultihashDigest trait

Split the `MultihashDigest` trait into the parts that need a hasher implementation
and those who don't. This way Multihashes can be created that don't have a specific
hasher associated.

vmx committed:
feat: add BasicMultihash

There are cases, where you don't care about a specific hasher, you just
want to parse/process a Multihash. For example if you encounter CIDs
that come from the outside that contain Multihashes that are not fully
supported by your application. This is what `BasicMultihash` is for.

dvc94ch commented on src/code.rs:
```diff
@@ -1,95 +1,82 @@
+    #[mh(code = 0xa0, hasher = crate::Strobe256)]
+    Strobe256(crate::StrobeDigest<crate::U32>),
+    /// Multihash array for hash function.
+    #[mh(code = 0xa1, hasher = crate::Strobe512)]
```
I guess you want to define the default codes as constants and use them here so:
`code = crate::STROBE_512`

dvc94ch reviewed (commented):
I think we can live without the code trait, I agree. I'm not completely convinced about the BasicMultihash, do we need that? Let me know once it's finished, I think it is looking good.

dvc94ch commented on src/multihash.rs:
```diff
@@ -47,38 +48,72 @@ pub trait MultihashDigest<C: MultihashCode>: Clone + Debug + Eq + Send + Sync +
+pub trait MultihashCreate: Clone + Debug + Eq + Send + Sync + 'static {
     /// Returns the hash of the input.
-    fn digest(&self, input: &[u8]) -> Self::Multihash;
+    fn new(code: u64, input: &[u8]) -> Result<Self, Error>;
```
Error type should probably be `UnsupportedCode`. That way most applications can just unwrap here.

vmx replied:
I don't understand your comment. The implementation currently returns an `Error::UnsupportedCode()`. Is there a way to specify more strictly in the return type?

dvc94ch replied:
Create a second error type

vmx replied:
I don't have enough experience with library APIs. So far I've mostly encountered one error type per library. But we can certainly introduce a new error type if that makes the API nicer.

dvc94ch replied:
Well if the only possible error is UnsupportedCode, but you return an enum, it's not clear that it's safe to unwrap without reading the code. If the error type is `struct UnsupportedCode`, it's clear that it can be unwrapped. The error enum can include the unsupported code error as #[from] UnsupportedCode, if it's necessary

dvc94ch replied:
I guess it's not that important, do you want to merge?

vmx replied:
Thanks for the elaborate reply, that makes sense now. Yes I'll merge.

vmx commented:
> I'm not completely convinced about the BasicMultihash, do we need that?

The use case for `BasicMultihash` (if someone has a better name, please let me know) is when you just want to parse an existing Multihash, but you don't really want to hash new data. E.g. you want to validate if a Multihash has the correct code value. You might not want to bundle a full hasher implementation for that.

dvc94ch commented:
I'm not sure how useful that is, I assume you want to figure out if the hash matches the data. Do you have a specific use case in mind?

vmx commented:
> Do you have a specific use case in mind?

One specific use case is Filecoin, where some very complex hashing is used (`poseidon-bls12_381-a2-fc1`), which you won't implement. You also wouldn't verify if the hash matches. Without `BasicMultihash` you couldn't parse such a Multihash.

Another case is that you could use it to deal with truncated hashes. If you get an error that the size doesn't match, you try using a `BasicMultihash` instead. It's basically for application that may encounter arbitrary multihashes.

dvc94ch commented:
Wouldn't the custom multihash table just use the identity hasher?

vmx commented:
> Wouldn't the custom multihash table just use the identity hasher?

That's a good point. It's an interesting hack. As I haven't worked on any code that would need the `BasicMultihash`, that's good enough that I'll remove it from this PR (and keep in in mind in case it turns out I'd need it).

This leads to: I did the split between `MultihashDigest` and `MultihashCreate` because of the `BasicMultihash`. Should I merge them again into a single trait?

vmx commented:
> Wouldn't the custom multihash table just use the identity hasher?

Oh wait. The problem is that with a custom Multihash table, you would still need to enumerate your supported hash coes. `BasicMultihash` works on just any code, without knowing it upfront (imagine a tool that lists all the different hashes used in a system that can store arbitrary IPLD data).

dvc94ch commented:
One problem with that is that the identity digest can only be used once, maybe the identity hasher can only be used once due to the IdentityDigest. The identity hasher should maybe be generic over the digest type.

> imagine a tool that lists all the different hashes used in a system that can store arbitrary IPLD data

RawMultihash might be a better name. I can see it's utility. The cid would use it presumably. So the custom Multihash should convert to and from RawMultihash and RawMultihash should convert to and from bytes?

vmx commented:
> I can see it's utility. The cid would use it presumably.

I had CIDs in mind, but I didn't get to rust-cid yet, to see how all this integrates :)

> So the custom Multihash should convert to and from RawMultihash

I'll add that.

> and RawMultihash should convert to and from bytes?

There is already the `to/from_bytes()` in the `MultihashDigest` trait. Do you mean that? Or do you mean a the `From/TryFrom` trait?

dvc94ch commented:
> There is already the to/from_bytes() in the MultihashDigest trait. Do you mean that? Or do you mean a the From/TryFrom trait?

Whatever you think makes sense

> I had CIDs in mind, but I didn't get to rust-cid yet, to see how all this integrates :)

I guess we should call it RawCid and make the Codec just a number like here.

vmx committed:
fix: rename `BasicMultihash` to `RawMultihash`

vmx committed:
refactor: clean up the proc macro a bit

vmx commented:
I just pushed minor changes, tomorrow I'll do one more round of cleanups.

> So the custom Multihash should convert to and from RawMultihash

This is harder than I expected. The problem is that you need to use the actual size of the digest stored in `RawMultihash` (and not the allocated one) (that's easy) and compare it to the digest size of the custom Multihash digests (this is hard).

I might skip that for now and look into how this all works together with rust-cid, rather then spending more time on it (and come back to it later).

I'll let you know when this PR is ready from my perspective (should be tomorrow).

vmx committed:
fix: better error messages

vmx committed:
chore: remove commented out code

vmx committed:
fix: use constants for multihash codes

vmx committed:
fix: enable missing_docs lint again

vmx committed:
test: enable arb again

vmx committed:
test: clean up tests a bit

vmx committed:
fix: fix clippy warnings

vmx commented:
@dvc94ch Now it's ready for the final review.

There are still a few things I'd like to fix, but I propose doing this in subsequent PRs. The current PR is good enough, it supports all the features (and more) that we already have.

Things I'd like to fix next are:

 - Return the actual size of the digest and the correct slice (this only matters for the Identity hash, where we don't know what the actual size is)
 - rename `code.rs` to `multihash_impl.rs` to make it clearer what it is about
   - also rename the `code` feature to something else, what about `multihash-impl` or `default-impl`
 - Add more documentation

dvc94ch reviewed (approved):
Looks good to me

dvc94ch commented:
Why is the maximum size for a digest 255 and not 32 or 64? I think you should make the maximum size 64 and use an array with length 64 instead of 32. at the moment you're only reading the first 32 bytes. event if you ignore the bytes you need to read till the end of the multihash.

`for i in 0..rest { read_exact(&mut [0]) }`

vmx commented:
> Why is the maximum size for a digest 255 and not 32 or 64?

I thought about the maximum which is possible with the current code. We use `ArrayLength<u8>` for a digest, hence I thought it's 255. But I may misunderstand what the `u8` in `ArrayLength` does.

dvc94ch commented:
Sorry missread the code, it looks like my comment doesn't make much sense.
